### PR TITLE
refactor(protocol-engine): simplify command models for usage in HTTP API schema

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This is a work-in-progress branch to simplify the Protocol Engine's command model so that it's easier to work with _and_ so it can be dropped into the public HTTP API schema without modification.

This is precursor work for #7870 and #7871

**Currently looking for feedback about how to cut the size of this PR down.**

## Changelog

TBD

## Review requests

TBD. Currently open questions:

- How much in this draft is really necessary?
- How do we route a command request to a full command model?

## Risk assessment

TBD

### CODE REVIEW

File: api/src/opentrons/protocol_engine/commands/add_labware_definition.py [hunk 1]
```diff
@@ -1,22 +1,23 @@
...
-class AddLabwareDefinitionRequest(BaseModel):
+
+class AddLabwareDefinitionData(BaseModel):
     """Request to add a labware definition."""
-    definition: LabwareDefinition = Field(
-        ...,
-        description="The labware definition."
-    )
 
-    def get_implementation(self) -> AddLabwareDefinitionImplementation:
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/commands/add_labware_definition.py:17]: Ditching `get_implementation` may be too much for this PR. If we keep `get_implementation`, we can punt on some of the harder questions, continue to get the models fundamentally simplified, and keep the command execution refactor more focused

File: api/src/opentrons/protocol_engine/clients/sync_client.py [hunk 2]
```diff
@@ -35,10 +35,12 @@ def load_labware(
     ) -> commands.LoadLabwareResult:
         """Execute a LoadLabwareRequest and return the result."""
         request = commands.LoadLabwareRequest(
-            location=location,
-            loadName=load_name,
-            namespace=namespace,
-            version=version,
+            data=commands.LoadLabwareData(
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/clients/sync_client.py:38]: This is awkward and kind of annoying, but in the absence of something like samuelcolvin/pydantic#717, this might just be what we have to do to get our HTTP model alignment

File: api/src/opentrons/protocol_engine/commands/base.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+class CommandStatus(str, Enum):
+    """Command execution status."""
+
+    QUEUED = "queued"
+    RUNNING = "running"
+    EXECUTED = "executed"
+    FAILED = "failed"
+
+
+class BaseCommandImpl(ABC, Generic[CommandDataT, CommandResultT]):
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/commands/base.py:31]: Part of me wants to go `AbstractCommand...` rather than`BaseCommand...` for all these

File: api/src/opentrons/protocol_engine/commands/command.py [hunk 4]
```diff
@@ -1,119 +0,0 @@
-"""Base command data model and type definitions."""
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/commands/command.py:1]: This file name is unhelpful, but maybe for this PR we can avoid renaming it to keep the diff smaller (and add a rename TODO instead)

File: api/src/opentrons/protocol_engine/commands/command_builder.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
+"""Command resource factory."""
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/commands/command_builder.py:1]: I don't really love this... Avoiding the removal of `get_implementation` it seeming more and more like the way to go

File: robot-server/robot_server/service/session/command_execution/__init__.py [hunk 6]
```diff
@@ -1,4 +1,9 @@
-from .base_command_queue import CommandQueue  # noqa: F401
+from .base_queue import CommandQueue  # noqa: F401
```
Comment (mcous) [inline robot-server/robot_server/service/session/command_execution/__init__.py:1]: This looks like a find-and-replace error

**Review by mcous — COMMENTED** [review]
TL;DR: I'm feeling like the command execution stuff (`get_implementation`) needs to remain stable for this PR, and we'll get a smaller, easier diff that way

File: api/src/opentrons/protocol_engine/commands/base.py (hunk 3 above)
Comment (sanni-t) [reply api/src/opentrons/protocol_engine/commands/base.py:31]: Can you remind me again why we picked `Base` over `Abstract` here? I think I remember suggesting `Base` for the ones below because they aren't ABCs and we don't expect them to be used the same way as the rest of our abstract classes. But this specific class is an ABC and has an abstract method. Did we decide to go with `Base..` here for uniformity purposes?

File: api/src/opentrons/protocol_engine/clients/sync_client.py (hunk 2 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocol_engine/clients/sync_client.py:38]: Notes from call:

* This extra level of nesting through `data` is not new. It's present today in the public HTTP API. This PR also brings it into the Protocol Engine code.
* For our internal code, we find this extra nesting a little bit annoying, but not too bad.
* For the public HTTP API, we (or at least I) think it's more of a bummer.
* I think we can use a custom `@root_validator` to eliminate it from the public HTTP API, as described in the Pydantic issue linked above.
    * I think this would also solve problems with bad discriminated union error messages, and with redundancy between `Union`s and inheritance hierarchies.
    * We think we can defer this change until after this PR. It won't incur too much churn.

**Comment by mcous** [comment]
Closing in favor of #7993
